src/common/valtext.cpp

```diff
@@ -215,114 +196,113 @@ bool wxTextValidator::TransferFromWindow()
-// wxStringCheck(). And while this might be fractionally less efficient because
-// the function call won't be inlined like this, we don't care enough about
-// this to add extra #ifs for non-IRIX case.
-namespace
-{
-
-bool CheckString(bool (*func)(const wxUniChar&), const wxString& str)
+wxString wxTextValidator::IsValid(const wxString& str) const
 {
-    for ( wxString::const_iterator i = str.begin(); i != str.end(); ++i )
+    if ( HasFlag(wxFILTER_EMPTY) && str.empty() )
+        return _("Required information entry is empty.");
+    else if ( IsExcluded(str) )
+        return wxString::Format(_("'%s' is one of the invalid strings"), str);
+    else if ( !IsIncluded(str) )
+        return wxString::Format(_("'%s' is not one of the valid strings"), str);
+
+    // check the whole string for invalid chars.
+    for ( wxString::const_iterator i = str.begin(), end = str.end();
+          i != end; ++i )
     {
-        if ( !func(*i) )
-            return false;
+        if ( !IsValid(*i) )
+        {
+            return wxString::Format(
+                _("'%s' contains invalid character(s)"), str);
+        }
     }
 
-    return true;
+    return wxString();
 }
 
-} // anonymous namespace
 
-wxString wxTextValidator::IsValid(const wxString& val) const
-{
-    // wxFILTER_EMPTY is checked for in wxTextValidator::Validate
-
-    if ( HasFlag(wxFILTER_ASCII) && !val.IsAscii() )
-        return _("'%s' should only contain ASCII characters.");
-    if ( HasFlag(wxFILTER_ALPHA) && !CheckString(wxIsalpha, val) )
-        return _("'%s' should only contain alphabetic characters.");
-    if ( HasFlag(wxFILTER_ALPHANUMERIC) && !CheckString(wxIsalnum, val) )
-        return _("'%s' should only contain alphabetic or numeric characters.");
-    if ( HasFlag(wxFILTER_DIGITS) && !CheckString(wxIsdigit, val) )
-        return _("'%s' should only contain digits.");
-    if ( HasFlag(wxFILTER_NUMERIC) && !wxIsNumeric(val) )
-        return _("'%s' should be numeric.");
-    if ( HasFlag(wxFILTER_INCLUDE_CHAR_LIST) && !ContainsOnlyIncludedCharacters(val) )
-        return _("'%s' doesn't consist only of valid characters");
-    if ( HasFlag(wxFILTER_EXCLUDE_CHAR_LIST) && ContainsExcludedCharacters(val) )
-        return _("'%s' contains illegal characters");
-
-    return wxEmptyString;
-}
-
-bool wxTextValidator::ContainsOnlyIncludedCharacters(const wxString& val) const
+void wxTextValidator::SetCharIncludes(const wxString& chars)
 {
-    for ( wxString::const_iterator i = val.begin(); i != val.end(); ++i )
-        if (m_includes.Index((wxString) *i) == wxNOT_FOUND)
-            // one character of 'val' is NOT present in m_includes...
-            return false;
+    for ( wxString::const_iterator i = chars.begin(), end = chars.end();
+          i != end; ++i )
+    {
+        if ( m_charExcludes.Find(*i) != wxNOT_FOUND )
+        {
+            wxLogWarning(_("A char should either be included or excluded!"));
```

Comment:
I t think this should be `wxFAIL_MSG()` and not a `wxLogWarning()`. The latter is meant for the user and what can a user do here? We really want to complain to the developer here, and this is what the asserts are for.